multy_core/src/bitcoin/bitcoin_transaction_segwit.cpp

```diff
@@ -0,0 +1,506 @@
+
+    BinaryDataPtr signature;
+    {
+        BitcoinDataStream sign_data;
+        sign_data << static_cast<int32_t>(2); // nVersion bip143 implimented
+        sign_data << as_binary_data(hash_prevouts);
+        sign_data << as_binary_data(hash_sequence);
+        sign_data << reverse(**prev_transaction_hash);
+        sign_data << prev_transaction_out_index;
+        sign_data << static_cast<uint8_t>(0x19);
+        sign_data << OP_DUP;
+        sign_data << OP_HASH160;
+        sign_data << as_compact_size(public_key_hash_data.len);
+        sign_data << public_key_hash_data;
+        sign_data << OP_EQUALVERIFY;
+        sign_data << OP_CHECKSIG;
+        sign_data << *amount;
+        sign_data << **sequence;
+        sign_data << as_binary_data(hash_outputs);
+        sign_data << *lock_time;
+        sign_data << static_cast<uint32_t>(1); //signature version
+
+        signature = private_key.get_value()->sign(sign_data.get_content());
+    }
+
+    BitcoinDataStream sig_script_witness_stream;
+    sig_script_witness_stream << uint8_t(2); // elements count  (signature, public key)
+    sig_script_witness_stream << as_compact_size(signature->len + 1);
+    sig_script_witness_stream << *signature;
+    sig_script_witness_stream << uint8_t(1); // hash-code type
+
+   // const PublicKeyPtr& public_key = private_key.get_value()->make_public_key();
+    const BinaryData& public_key_data = public_key->get_content();
+    sig_script_witness_stream << as_compact_size(public_key_data.len);
+    sig_script_witness_stream << public_key_data;
+
+    script_witness = make_clone(sig_script_witness_stream.get_content());
+}
+
+BinaryDataPtr BitcoinTransactionSegWitSource::serialize_with_script_pub_key() const
+{
+    BitcoinDataStream result;
+
+    result << reverse(**prev_transaction_hash);
+    result << prev_transaction_out_index;
+
+    BinaryDataPtr script_pubkey(make_script_pubkey());
+    result << as_compact_size(script_pubkey->len);
+    result << *script_pubkey;
+    result << **sequence;
+
+    return make_clone(result.get_content());
+}
+
+BinaryDataPtr BitcoinTransactionSegWitSource::preview_transaction() const
+{
+    BitcoinDataStream result;
+
+    result << reverse(**prev_transaction_hash);
+    result << prev_transaction_out_index;
+
+    return make_clone(result.get_content());
+}
+
+BinaryDataPtr BitcoinTransactionSegWitSource::make_script_pubkey() const
+{
+    BitcoinDataStream result;
+
+    const PublicKeyPtr& public_key = private_key.get_value()->make_public_key();
+    std::array<uint8_t, HASH160_LEN + 2> segwit_script_P2WPKH;
+    BinaryData public_key_hash_data = power_slice(as_binary_data(segwit_script_P2WPKH), 2, -2);
+    segwit_script_P2WPKH[0] = OP_VERSION_BYTE_WITNESS; // Version byte witness
+    segwit_script_P2WPKH[1] = HASH160_LEN; // Witness program is 20 bytes
+    bitcoin_hash_160(public_key->get_content(), &public_key_hash_data);
+
+    result << static_cast<uint8_t>(0x16); // size script to P2SH
```

Comment:
wth ? isn't that supposed to be `segwit_script_P2WPKH.size()` or something similar?